src/webview/html/messageAsHtml.js

```diff
@@ -71,13 +75,72 @@ $!${messageReactionListAsHtml(reactions, ownUser.user_id, allImageEmojiById)}
 `;
 };
 
-const widgetBody = (message: Message | Outbox) => template`
+const widgetBody = (message: Message, ownUserId: UserId) => {
+  const incompatible_message = template`
 $!${message.content}
-<div class="special-message"
- ><p>Interactive message</p
- ><p>To use, open on web or desktop</p
-></div>
+<div class="special-message">
+  <p>Interactive message</p>
+  <p>To use, open on web or desktop</p>
+</div>
 `;
+  if (!message.submessages) {
+    return incompatible_message;
+  }
+
+  const widgets = message.submessages
+    .filter(submessage => submessage.msg_type === 'widget')
+    .map(submessage => ({
+      ...submessage,
+      content: JSON.parse(submessage.content),
+    }));
+
+  const poll_widgets = widgets.filter(widget => widget.content.widget_type === 'poll');
+  const events = widgets.filter(widget => 'type' in widget.content);
```

Comment:
The [widget](https://zulip.readthedocs.io/en/latest/subsystems/widgets.html) docs are the closes we have to docs for this, but it won't answer your question. [`PollData`](https://github.com/zulip/zulip/blob/master/static/shared/js/poll_data.js) is where the code for this lives, which is the closest thing we have to documentation.

I'll go add some comments to that file about the format of what these messages look like. In order to develop this, I just looked at the messages from the API in order to develop this, since there weren't any docs.

[Here's](https://gist.github.com/WesleyAC/7cd8e7a94ecd597d739197b83949171e) the message data for [this message](https://chat.zulip.org/#narrow/stream/101-design/topic/bottom.20row.20design/near/1170377), in case that helps you review.

Reply:
"events" is I guess a term that I made up for the submessages that aren't the initial widget submessage (`vote` and `new_option`, and a few others), I'll add a comment explaining that.

Reply:
Added a comment about this term.